### PR TITLE
feat: Add support for encrypted NCSD

### PR DESCRIPTION
**Pull request by marysaka**
This adds support for encrypted NCSD files.

We loads keys from `<data_dir>/Alber/sysdata/aes_keys.txt`.

NOTE: We do differ from other emulators by not hard-coding the `generator` key, it's the user responsibility to provide it in `aes_keys.txt`.

NOTE: This is missing NCCH version 1 support and seeded crypto but that would be enough for now.

### CODE REVIEW

File: include/crypto/aes_engine.hpp
Comment (wheremyfoodat) [inline include/crypto/aes_engine.hpp:32]: Can the uses of UINT8_WIDTH here be replaced with `constexpr size_t u8_width = 8;` UINT8_WIDTH is an optional C23 addon

File: src/core/loader/ncch.cpp
Comment (wheremyfoodat) [inline src/core/loader/ncch.cpp:121]: *truly :D

**Comment by wheremyfoodat** [comment]
I don't see anything wrong and I trust you with crypto stuff but it would be nice to get a second review from someone else.
Has this been tested on multiple encrypted/unencrypted games?

**Comment by wheremyfoodat** [comment]
Probably want to make the AES CTR key optionals (keyX, keyX, normalKey, encryptionInfo) explicitly initialized to std::nullopt

**Comment by marysaka** [comment]
> I don't see anything wrong and I trust you with crypto stuff but it would be nice to get a second review from someone else. Has this been tested on multiple encrypted/unencrypted games?

I tested it with around 10 different encrypted games and then with decrypted form.

**Review by Wunkolo — APPROVED** [review]
panda

File: include/crypto/aes_engine.hpp
Comment (Wunkolo) [inline include/crypto/aes_engine.hpp:13]: `std::size_t` -> `usize`
`uint8_t` -> `u8`
etc

File: include/crypto/aes_engine.hpp
Comment (Wunkolo) [inline include/crypto/aes_engine.hpp:18]: All of these `const std::array<uint8_t, N>&` can be `std::span<const u8, N>` instead so it can take any container of randomly-addressable continuous elements.

File: src/emulator.cpp
Comment (Wunkolo) [inline src/emulator.cpp:143]: This can throw an exception in the case of an OS-level error. Might want to use the `std::error_code` form of this function instead to explicitly report them.
https://en.cppreference.com/w/cpp/filesystem/exists

**Review by Wunkolo — CHANGES_REQUESTED** [review]
just some nit picks

File: include/crypto/aes_engine.hpp
Comment (wheremyfoodat) [reply include/crypto/aes_engine.hpp:13]: I changed the uint8_t to u8 but the codebase doesn't use the `usize` type in helpers.hpp that much atm.
Maybe we should have a poll among contributors about it?

File: include/crypto/aes_engine.hpp
Comment (wheremyfoodat) [reply include/crypto/aes_engine.hpp]: The std::span method actually is likely less intuitive here actually, I think

```cc
constexpr void updateNormalKey(std::size_t slotId) {
			if (m_generator.has_value() && hasKeyX(slotId) && hasKeyY(slotId)) {
				auto &keySlot = m_slots.at(slotId);
				AESKey keyX = keySlot.keyX.value();
				AESKey keyY = keySlot.keyY.value();

				keySlot.normalKey = rolArray(addArray(xorArray(rolArray(keyX, 2), keyY), m_generator.value()), 87);
			}
		}
		```
		The last line has a sort of snowball to process the array which works a lot better with arrays vs spans imo
		I think we should keep it, at least for now.

**Pull request merged**
